### DISCUSSION TITLE
Users receive messages/notifications based on their progress on Prashnavali

### DISCUSSION THREAD

**Discussion by tarang27** [Sunbird-Lern/Community#31] [Query]
Hi All,

Sharing a brief note on the new feature that we have planned as part of Prashnavali solution on VDN for the 5.0 release.

Functionality:

The admin is able to manually send actionable progress based messages to creators & reviewers

Challenges in the current solution flow:

There are 3 overarching use cases for which Prashnavali is being used for creation of good quality questions:-

Assessments
Practice Worksheets
Question Banks
The mentioned use cases require a quick turnaround time and close monitoring on the progress of the status and regular reminders to users to complete the assigned task. Just to add some colour:-

For an assessment, usually 5 days are allotted to the creator & contribution reviewer to create, review, make required changes, resubmit and re-review around 40 questions each.

For a practice worksheet, each creator has to complete two chapters per week (15 questions each), and each reviewer has to review questions for around 4 chapters per week (15 questions each)

Currently, to send reminders to each creator & reviewer, the admin has to do these steps:-

Login to the contribution portal
Go to ‘My Projects’
Open the desired project (approx. 46 projects in chapter worksheets)
See the overall progress status of the project
Open the desired question set (approx 15-30 question sets per project in case of chapter worksheets)
See the progress in that particular question set
Send a reminder/nudge to the creator & reviewer
Steps 3-7 need to be repeated for all projects & all question sets in that project

Functional solution:

The admin is able to send reminders/nudges to users (creators & reviewers) based on their progress in a project

A detailed PRD has also been created for this here:- https://project-sunbird.atlassian.net/wiki/spaces/PRD/pages/3197403157/Users+receive+messages+based+on+their+progress+on+Prashnavali

Would request you to go through it and share your comments. Please help us understand the best way to implement this feature.
@surendrasinghs @pallakartheekreddy @rhwarrier @reshmi-nair

**Comment by reshmi-nair** [comment]
@tarang27 In sunbird, there are 3 types of notification - In-App notifications(User Feeds), Email and SMS. These notifications can be send to specific user using Notification Service in Lern BB. 
Suggesting to checkout In-App user notifications for above functionality, as these can be fetched using notification feed read APIs by passing user identifier to display after user login.

**Reply by tarang27** [reply]
Hi @reshmi-nair thanks for the quick turnaround.
A few queries, 
For in-app notifications:
1. Would the user be required to have logged in to the portal to see those notifications ?
2. How would the notifications be displayed in case the user is using:- a.) the web portal b.) mobile application

In case the answer for 1 is yes, we would not prefer going with that approach, since the idea is, that nudges to the users are sent to have them log-in to the portal and complete the task assigned to them.

**Comment by rhwarrier** [comment]
Yes, only logged in users get to see in-app notifications
The user sees an indication in the menu - showing they have a notification. 

A few questions worth answering in this scenario are:
1. Are the target group of users registered with phone or email?
2. Are they likely to see the nudges, given there is a lot of SMS spam these days, and most people rarely check email on a regular basis?
3. Do you have a way to measure if these nudges, when sent out, are achieving their purpose?
4. Is it possible to achieve this outcome outside of the system? Most of such employees have active WhatsApp groups that are used for sharing info - would it be possible to leverage one of those streams, as it would be more likely to get noticed, and hence acted upon?

**Reply by tarang27** [reply]
Hi @rhwarrier ,

Thanks for your response on this, it is very helpful.

So currently, this entire use case is being managed outside the system via manual tracking of project level progress, manual tracking of user level progress & then sending nudges on whatsapp groups which is very error prone and cumbersome. The detailed user actions are mentioned in the PRD.

We are wanting to automate this process as much as possible wherein a admin controlled & system generated SMS is shared to the contributor/reviewer based on their progress. 

For this, needed your help to understand how can the tech solution be designed, what all APIs need to be called ?
Also would be helpful to know your thoughts on how to implement an MVP version of this use case ?

**Reply by rhwarrier** [reply]
Automating an inefficient process may not necessarily make it efficient (at least, based on some of our experiences) - It may be necessary to think about it ground up, and understand what will work best, given user behaviour. Which is why some of these points were raised. 
Your team will be best placed to think up workflows - would request to raise this as an RFC on the discussion forum once you decide on the approach, so that the community can also chip in with thoughts. 
Another Sunbird BB worth exploring in this context could be Sunbird UCI - https://uci.sunbird.org
Please connect with the team if you need to explore further on whether this can help with your needs. Thanks 

@vrayulu @SanthoshVasabhaktula

**Reply by sukhpreetssekhon** [reply]
@tarang27 - do see if Sunbird UCI can be leveraged for this. Will be good to understand the delta required in UCI to make this work. 

cc - @chakshugautam

**Reply by tarang27** [reply]
Thanks @rhwarrier @sukhpreetsamagra,

We will look into the usage of Sunbird UCI for this use case.

@rhwarrier from our understanding, a similar implementation is already in place. Under the question set flow on Vidyadaan, whenever a reviewer rejects a question, the contributor of that question gets an SMS notifying the rejection of that question. Wanted to understand a few things on that implementation:-

1. What APIs are being used for this service ?
2. How is the SMS being sent to the contributor ? (any PRD would be helpful) 
3. Which SMS provider is being used for this service ?
4. Who bears the cost for this service ?

It would be very helpful to know the answers to the above for initiating exploration of the method of implementation.

Thanks

**Comment by rhwarrier** [comment]
@surendrasinghs  @pallakartheekreddy  could you please help with the details of the VDN flow? Please see questions about. Thanks.

**Reply by tarang27** [reply]
Hi @surendrasinghs, can you please help with this ?

**Comment by surendrasinghs** [comment]
@tarang27 Please see responses in-line
> 
> from our understanding, a similar implementation is already in place. Under the question set flow on Vidyadaan, whenever a reviewer rejects a question, the contributor of that question gets an SMS notifying the rejection of that question. Wanted to understand a few things on that implementation:-
> 
> 1. What APIs are being used for this service ?

Which question set flow are you referring to - QuML 0.5 or QuML 1.0? Is this really implemented? 
When we inspect the question set flow it is not calling Notification service API. There is no notification service API that gets called for question publish/ reject. (see screenshot). 
Notification service is being used for this. Refer to https://lern.sunbird.org/learn/product-and-developer-guide/notification-service
![image](https://user-images.githubusercontent.com/10580754/187891050-cdece6bd-4fd1-4996-b914-9c84cc5ef0b2.png)


> 2. How is the SMS being sent to the contributor ? (any PRD would be helpful)

Email/SMS Notification for Content workflow was implemented around May 2020. Refer to https://project-sunbird.atlassian.net/browse/DP-11 

> 3. Which SMS provider is being used for this service ?

Specific adopters of open-source building blocks can configure SMS providers as per their region and choice. For example, msg91 and SendGrid could be the choice of the service providers in India.


> 4. Who bears the cost for this service ?

The adopter of the open-source building block.

> 
> It would be very helpful to know the answers to the above for initiating exploration of the method of implementation.
> 
> Thanks

**Reply by tarang27** [reply]
Thanks for the response @surendrasinghs , this is very helpful.
We would want to implement something similar to [Link](https://project-sunbird.atlassian.net/browse/DP-11) for the Question set flow.
